ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/worldstate/Pruner.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.worldstate;
+
+import tech.pegasys.pantheon.ethereum.ProtocolContext;
+import tech.pegasys.pantheon.ethereum.chain.BlockAddedEvent;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.core.Hash;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class Pruner {
+  private static final Logger LOG = LogManager.getLogger();
+  private final MarkSweepPruner pruningStrategy;
+  private final ProtocolContext<?> protocolContext;
+  private final ExecutorService executor;
+  private final long retentionPeriodInBlocks;
+  private final AtomicReference<State> state = new AtomicReference<>(State.IDLE);
+  private volatile long markedBlockNumber = 0;
```

Comment:
Why does this need to be volatile?

Reply:
Because new block notifications could be arriving from multiple threads.